### PR TITLE
[FEAT] Auth Service

### LINKED ISSUE #62: [FEAT] Auth Service
# Contents

인증을 담당하는 Auth Service와 신원 정보를 담당하는 Credential Service를 구현한다.

## Description

- AuthService: 로그인/로그아웃 및 토큰 발급을 담당한다. refresh token은 device별로 저장되어 관리된다. 최대 5개까지 
   - login
   - issue: OAuth login 및 Signup에서 토큰만 발행하고자할 때 사용한다
   - reissue
   - logout
- CredentialSerivce: AuthService, UserService 등에서 신원을 확인할 정보를 필요로 할 때 사용한다.
   - getCredential: id, email, pw 등 인증에 필요한 정보들을 가져온다.
   - updatePassword
   - isPasswordMatch
- TokenOutputPort, CredentialOutputPort

## Todo

-   [ ] AuthService
   - [ ] device 최대 5기기 설정
-   [ ] AuthServiceT
-   [ ] JwtProvider
-   [ ] CredentailService
-   [ ] CredentialServiceT
-   [ ] TokenOutputPort
-   [ ] CredentialOutputPort

## Check Priority

Expected Priority: HIGH

## Reference (Optional)

(null)

### PR DESCRIPTION
**Pull request by sueelly**
## #️⃣ Related Issue

Linked Feature backlog: close #62 

## 📝 Purpose of PR

JwtProvider, AuthService와 CredentialService를 구현한다.

## Contents

- AuthService는 사용자 인증과 토큰 발급을 담당한다.
- CredentialService는 사용자의 credential 데이터를 다룬다. 비밀번호 변경이 이에 해당된다

### sign in logic

1. AuthSerivce signIn(email, password, deviceId)
2. CredentialService isPasswordMatch(email, password)
3. CredentialService getCredential(email)
4. PasswordService isPasswordMatch(password, credential.getPassword())
5. JwtProvider createToken(userId, deviceId)
6. DeviceCredentialService login(userId, refreshToken)
7. return token

credential을 조회해야만 비밀번호 확인이 가능하다. 따라서 비밀번호 확인 시 CredentialService를 사용하고, CredentialService에서 PasswordService를 호출하여 비밀번호 로직을 수행한다.
로그인 정보는 각 기기별로 저장되며, 해당 정보를 다루는 곳이 DeviceCredentialService이다.
- reissue 시 로그인 상태를 확인하고, refresh token을 업데이트한다.
- logout 시 로그인 상태를 업데이트하고, refresh token을 제거한다.

**singIn sequential diagram**
singIn

## Checklist

-   [x] Does commit messages follow the convention?
-   [x] Are variable names brief but descriptive?
-   [x] Is the code clean and easy to understand?
-   [x] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
-   [ ] Does the test successfully test edge and corner cases?

## 💬 Review Requirements (Optional)

Is there anything you can praise about this PR? Start with praise.
Make a review, leaving kind comments and suggesting changes where needed (to resolve the above).

### CODE REVIEW

File: snapnow-application/src/main/java/kr/flab/snapnow/application/auth/jwt/TokenPayload.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+        TokenPayload that = (TokenPayload) obj;
+        return userId.equals(that.userId) && deviceId.equals(that.deviceId)
+                && issuedAt.equals(that.issuedAt);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(userId, deviceId, issuedAt);
+    }
```
Comment (f-lab-jason) [inline snapnow-application/src/main/java/kr/flab/snapnow/application/auth/jwt/TokenPayload.java:47]: token 레벨에서 이 코드들은 왜 필요한가요?

Comment (sueelly) [reply snapnow-application/src/main/java/kr/flab/snapnow/application/auth/jwt/TokenPayload.java]: 원래는 테스트에서 생성하기 전 만든 것과 생성 후에 만든 것이 같은지 비교하기 위해서 생성되었으나, 현재는 죽은 코드네요...
삭제토록 하겠습니다.
